src/sentry/api/endpoints/project_rules.py

```diff
@@ -44,66 +48,219 @@ def pre_save_rule(instance, sender, *args, **kwargs):
+        match_results.has_key = True
+        # Only compare if both have the key
+        if existing_actions and current_actions:
+            match_results.key_matches = self._compare_lists_of_dicts(
+                keys_to_ignore=["uuid"], list1=existing_actions, list2=current_actions
+            )
+
+        return match_results
+
+    @classmethod
+    def _compare_lists_of_dicts(
+        cls,
+        keys_to_ignore: list[str],
+        list1: list[dict[Any, Any]] | None = None,
+        list2: list[dict[Any, Any]] | None = None,
+    ) -> bool:
+        if list1 is None or list2 is None:
+            return False
+
+        if len(list1) != len(list2):
+            return False
+
+        for i, left in enumerate(list1):
+            right = list2[i]
+            raw_left = {k: v for k, v in left.items() if k not in keys_to_ignore}
+            raw_right = {k: v for k, v in right.items() if k not in keys_to_ignore}
+
+            # TODO (Yash): This code commented below is the corrected logic which accounts for bad key values.
+            # clean_left = cls._get_clean_actions_dict(raw_left)
+            # clean_right = cls._get_clean_actions_dict(raw_right)
+            # if clean_left != clean_right:
+            #     return False
+            """
+            This is a bug in the current logic.
+            When comparing DB values to serialized values, the values that are `None` are not properly converted to
+            empty strings.
+            This means we end up incorrectly evaluating the actions aren't the same, when they actually are.
+            """
+            if raw_left != raw_right:
+                return False
+
+        return True
+
+    @classmethod
+    def _get_clean_actions_dict(cls, actions_dict: dict[Any, Any]) -> dict[Any, Any]:
+        """
+        Returns a dictionary where None is substituted with empty string to help compare DB values vs serialized values
+        """
+        cleaned_dict = {}
+        for k, v in actions_dict.items():
+            cleaned_dict[k] = "" if v is None else v
+
+        return cleaned_dict
+
+    def find_duplicate(self) -> Rule | None:
+        """
+        Determines whether specified rule already exists, and if it does, returns it.
+        """
+        existing_rules = Rule.objects.exclude(id=self._rule_id).filter(
+            project__id=self._project_id, status=ObjectStatus.ACTIVE
+        )
+        for existing_rule in existing_rules:
+            keys_checked = 0
+            keys_matched = 0
+            for key_to_check in self._keys_to_check:
+                func = self._get_func_to_call(key_to_check=key_to_check)
+                results: MatcherResult = func(
+                    existing_rule=existing_rule, key_to_check=key_to_check
+                )
+                if results.has_key:
+                    keys_checked += 1
+                    if results.key_matches:
+                        keys_matched += 1
```

Comment:
After getting rid of has_key MatcherResult could just be a boolean saying if the self_func_to_call has found a full match. What I'm suggesting is basically you have one boolean which indicates `if both sides of the comparison have the key and they matched or none of them have it`

Reply:
Simply returning True/False isn't going to cut it, we have to be able to differentiate between `False` when the key is found, and `False` when the key is not found. We'd have to return `None` when the key is not found, `False` when the key is found but it doesn't match, and `True` when the key is found and it also matches. At that stage, we're assigning a special state for None. I can make that change in a follow up and see how the change looks

Reply:
what I meant was that `None` when both sides don't include the key and `True` for each key are the same state. Means the keys are equal. 

Can you explain on product behavior why should we differentiate when one side has they key and the other doesn't with when both have the key and not a match?

Reply:
We'll figure out what to do with the production direction in a follow up ticket: https://github.com/orgs/getsentry/projects/183/views/1?pane=issue&itemId=53027179

Right now we're just trying to unblock the UUID portion which is needed for Slack Thread support